### PR TITLE
Mhd/refactor compass action

### LINKED ISSUE #25: SceneView looses pitch if a compass is pressed
### Issue:
When tapping the Compass to rotate a SceneView back to North, any camera settings, such as pitch or roll, are lost. The issue was realized during implementation of the new compass component.

https://user-images.githubusercontent.com/16397058/153255099-b3bb07b9-7b06-4f09-a0f1-bf4e486d37fc.mp4

### Done Criteria:
Tapping the compass in a SceneView where pitch or roll have been applied will not result in these values being lost.

### Steps to Reproduce:

Add a Compass to a SceneView.  Rotate the scene and change the pitch, then tap the Compass (to rotate back to North).  The pitch setting has been reset.

### Area of Interest:
In `CompassViewModel.swift` `func resetHeading()` is responsible for reoreinting the viewpoint to north. This would presumably be the method for any compass related changes to support resolving this issue.

cc @mhdostal

### PR DESCRIPTION
**Pull request by mhdostal**
Closes #25, #266

This takes #281 and updates it a bit:

- Add `automaticallyHides` view modifier and removes `autoHide` from the constructors
- Removing `autoHide` allows the `action` to be a closure
- Adds `Viewpoint.withViewpoint(rotation:)` extension for ease of viewpoint creation when only the rotation changed.
- updates the MapView compass example to animate taps on the compass (rotation to 0).
- 
it includes these changes from the original PR:

- Adds an optional `action` parameter to the Compass initializer(s) that takes an async task. This allows users to perform animated heading changes. If no `action` is provided, behavior remains unchanged.
- The name "`action`" is borrowed from [`Button.init(action:label:)`](https://developer.apple.com/documentation/swiftui/button/init(action:label:))

### CODE REVIEW

File: Examples/Examples/CompassExampleView.swift [hunk 1]
```diff
@@ -15,27 +15,118 @@ import ArcGIS
...
+    
+    var body: some View {
+        SceneView(scene: scene, cameraController: cameraController)
+            .onCameraChanged { newCamera in
+                heading = newCamera.heading.rounded()
+            }
+            .overlay(alignment: .topTrailing) {
+                Compass(viewpointRotation: $heading) {
+                    // Animate the scene view when the compass is tapped.
+                    _ = try? await cameraController.moveCamera(
```
Comment (dfeinzimer) [inline Examples/Examples/CompassExampleView.swift:112]: Based on the team's answer to my question hopefully `_ =` can be removed

File: Documentation/Compass/README.md [hunk 2]
```diff
@@ -23,9 +23,10 @@ Compass:
     /// direction toward true East, etc.).
     /// - Parameters:
     ///   - heading: The heading of the compass.
+    ///   - action: An action to perform when the compass is tapped.
     ///   - autoHide: A Boolean value that determines whether the compass
```
Comment (dfeinzimer) [inline Documentation/Compass/README.md:27]:
```suggestion
```

File: Documentation/Compass/README.md [hunk 3]
```diff
@@ -35,18 +36,20 @@ Compass:
     /// - Parameters:
     ///   - viewpointRotation: The viewpoint rotation whose value determines the
     ///   heading of the compass.
+    ///   - action: An action to perform when the compass is tapped.
     ///   - autoHide: A Boolean value that determines whether the compass
```
Comment (dfeinzimer) [inline Documentation/Compass/README.md:40]:
```suggestion
```

File: Documentation/Compass/README.md [hunk 4]
```diff
@@ -35,18 +36,20 @@ Compass:
...
-    public init(viewpointRotation: Binding<Double>, autoHide: Bool = true)
+    public init(viewpointRotation: Binding<Double>, action: (() async -> Void)? = nil, autoHide: Bool = true)
 ```
 
 ```swift
     /// Creates a compass with a binding to an optional viewpoint.
     /// - Parameters:
     ///   - viewpoint: The viewpoint whose rotation determines the heading of the compass.
+    ///   - action: An action to perform when the compass is tapped.
     ///   - autoHide: A Boolean value that determines whether the compass automatically hides itself
```
Comment (dfeinzimer) [inline Documentation/Compass/README.md:50]:
```suggestion
```

**Comment by dfeinzimer** [comment]
It looks like removing the auto hide parameter broke a number of the tests.

Here's a prototyped fix for `testHiddenWithAutoHideOff()`:

```swift
func testHiddenWithAutoHideOff() {
    let initialValue = 0.0
    let finalValue = 90.0
    var _viewpoint: Viewpoint? = makeViewpoint(rotation: initialValue)
    let viewpoint = Binding(get: { _viewpoint }, set: { _viewpoint = $0 })
    let compass = Compass(viewpoint: viewpoint).automaticallyHides(false)
    XCTAssertFalse((compass as? Compass)!.shouldHide)
    _viewpoint = makeViewpoint(rotation: finalValue)
    XCTAssertFalse((compass as? Compass)!.shouldHide)
}
```

File: Sources/ArcGISToolkit/Extensions/Viewpoint.swift [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+                rotation: rotation
+            )
+        case.boundingGeometry:
+            return Viewpoint(
+                boundingGeometry: self.targetGeometry,
+                rotation: rotation
+            )
+        }
+    }
+}
```
Comment (dfeinzimer) [inline Sources/ArcGISToolkit/Extensions/Viewpoint.swift:35]: This strikes me as the type of thing we should add a test for, what do you think? I prototyped two quick tests that should hit both cases:

```swift
import ArcGIS
import XCTest

final class ViewpointTests: XCTestCase {
    func testWithRotationWithBoundingGeometry() {
        let rotatedViewpoint = Viewpoint(boundingGeometry: Polyline.simple.extent, rotation: -45)
        XCTAssertEqual(rotatedViewpoint.rotation, -45)
        
        let nonRotatedViewpoint = rotatedViewpoint.withRotation(.zero)
        XCTAssertEqual(nonRotatedViewpoint.rotation, 0)
    }
    
    func testWithRotationWithCenterAndScale() {
        let rotatedViewpoint = Viewpoint(center: .esriRedlands, scale: 1_000, rotation: 45)
        XCTAssertEqual(rotatedViewpoint.rotation, 45)
        
        let nonRotatedViewpoint = rotatedViewpoint.withRotation(.zero)
        XCTAssertEqual(nonRotatedViewpoint.rotation, 0)
    }
}

private extension Point {
    static var esriRedlands: Point {
        .init(
            x: -117.19494,
            y: 34.05723,
            spatialReference: .wgs84
        )
    }
}

private extension Polyline {
    static var simple: Polyline {
        .init(points: [
            Point(x: -117, y: 34, spatialReference: .wgs84),
            Point(x: -116, y: 34, spatialReference: .wgs84),
            Point(x: -116, y: 33, spatialReference: .wgs84)
        ])
    }
}
```

Comment (mhdostal) [reply Sources/ArcGISToolkit/Extensions/Viewpoint.swift:35]: Yes, those look good, thanks for the suggestion!

File: Sources/ArcGISToolkit/Components/Compass/Compass.swift [hunk 6]
```diff
@@ -125,4 +132,13 @@ public extension Compass {
         copy.size = size
         return copy
     }
+    
+    /// Specifies whether the ``Compass`` should automatically hide when the heading is 0.
+    /// - Parameter newAutomaticallyHides: A Boolean value indicating whether  the compass should automatically
+    /// hide/show itself when the heading is `0`.
+    func automaticallyHides(_ newAutomaticallyHides: Bool) -> some View {
```
Comment (dfeinzimer) [inline Sources/ArcGISToolkit/Components/Compass/Compass.swift:139]:
```suggestion
    func automaticallyHides(_ flag: Bool) -> some View {
```

After some thought I think `automaticallyHides` is a good name, how about the parameter's name though?

Apple uses [`flag`](https://developer.apple.com/documentation/familycontrols/familyactivitypicker/allowstightening(_:)#:~:text=Parameters-,flag,value%20that%20indicates%20whether%20the%20space%20between%20characters%20compresses%20when%20necessary.,-Return%20Value) for a similar parameter to a modifier.

File: Documentation/Compass/README.md [hunk 7]
```diff
@@ -35,27 +35,26 @@ Compass:
...
-    public init(viewpoint: Binding<Viewpoint?>, autoHide: Bool = true)
+    ///   - action: An action to perform when the compass is tapped.
+    public init(viewpoint: Binding<Viewpoint?>, action: (() async -> Void)? = nil)
 ```
 
-`Compass` has the following modifier:
+`Compass` has the following modifiers:
 
 - `func compassSize(size: CGFloat)` - The size of the `Compass`, specifying both the width and height of the compass.
+- `func automaticallyHides(newAutomaticallyHides: Bool)` - Specifies whether the ``Compass`` should automatically hide when the heading is 0.
```
Comment (dfeinzimer) [inline Documentation/Compass/README.md:53]: Copy Qualified Symbol Name produces this:
```suggestion
- `func automaticallyHides(_:)` - Specifies whether the ``Compass`` should automatically hide when the heading is 0.
```

**Review by dfeinzimer — COMMENTED** [review]
Just the one change and this should be good to go.

Up to you if you want to wait for the `moveCamera` PR to merge or not.

File: Sources/ArcGISToolkit/Components/Compass/Compass.swift [hunk 8]
```diff
@@ -125,4 +132,13 @@ public extension Compass {
         copy.size = size
         return copy
     }
+    
+    /// Specifies whether the ``Compass`` should automatically hide when the heading is 0.
+    /// - Parameter flag: A Boolean value indicating whether  the compass should automatically
```
Comment (dfeinzimer) [inline Sources/ArcGISToolkit/Components/Compass/Compass.swift:137]:
```suggestion
    /// - Parameter flag: A Boolean value indicating whether the compass should automatically
```

File: Examples/Examples/CompassExampleView.swift (hunk 1 above)
Comment (dfeinzimer) [reply Examples/Examples/CompassExampleView.swift:111]:
```suggestion
                    await cameraController.moveCamera(
```

File: Sources/ArcGISToolkit/Components/Compass/Compass.swift [hunk 9]
```diff
@@ -17,21 +17,24 @@ import SwiftUI
...
 public struct Compass: View {
-    /// A Boolean value indicating whether  the compass should automatically
-    /// hide/show itself when the heading is `0`.
-    private let autoHide: Bool
-    
     /// The opacity of the compass.
     @State private var opacity: Double = .zero
     
+    /// An action to perform when the compass is tapped.
+    private let action: (() async -> Void)?
```
Comment (philium) [inline Sources/ArcGISToolkit/Components/Compass/Compass.swift:24]: A user will likely want to animate the viewpoint (an asynchronous operation), but they may also want to do other things non-asynchronously. By making the action handler async, you've taken that control out of the users hands. If it were non-async, they could easily perform an async operation by wrapping it in a task (`Task { ... }`), just like this component is doing. But as an async closure, there is no way for them to perform an action inside the closure non-asynchronously. Consider making this non-async, just like a button action handler is.

Comment (mhdostal) [reply Sources/ArcGISToolkit/Components/Compass/Compass.swift]: Thanks, I made the change.

**Comment by mhdostal** [comment]
> Up to you if you want to wait for the moveCamera PR to merge or not.

No, let's get this merged, it's been pending a while. Making the follow-up change can happen with other changes as well.

File: Documentation/Compass/README.md [hunk 10]
```diff
@@ -23,9 +24,8 @@ Compass:
     /// direction toward true East, etc.).
     /// - Parameters:
     ///   - heading: The heading of the compass.
-    ///   - autoHide: A Boolean value that determines whether the compass
-    ///   automatically hides itself when the heading is `0`.
-    public init(heading: Binding<Double>, autoHide: Bool = true)
+    ///   - action: An action to perform when the compass is tapped.
+    public init(heading: Binding<Double>, action: (() async -> Void)? = nil)
```
Comment (dfeinzimer) [inline Documentation/Compass/README.md:28]:
```suggestion
    public init(heading: Binding<Double>, action: (() -> Void)? = nil)
```

File: Documentation/Compass/README.md [hunk 11]
```diff
@@ -35,27 +35,26 @@ Compass:
     /// - Parameters:
     ///   - viewpointRotation: The viewpoint rotation whose value determines the
     ///   heading of the compass.
-    ///   - autoHide: A Boolean value that determines whether the compass
-    ///   automatically hides itself when the viewpoint rotation is 0 degrees.
-    public init(viewpointRotation: Binding<Double>, autoHide: Bool = true)
+    ///   - action: An action to perform when the compass is tapped.
+    public init(viewpointRotation: Binding<Double>, action: (() async -> Void)? = nil)
```
Comment (dfeinzimer) [inline Documentation/Compass/README.md:39]:
```suggestion
    public init(viewpointRotation: Binding<Double>, action: (() -> Void)? = nil)
```

File: Documentation/Compass/README.md [hunk 12]
```diff
@@ -35,27 +35,26 @@ Compass:
...
 
 ```swift
     /// Creates a compass with a binding to an optional viewpoint.
     /// - Parameters:
     ///   - viewpoint: The viewpoint whose rotation determines the heading of the compass.
-    ///   - autoHide: A Boolean value that determines whether the compass automatically hides itself
-    ///   when the viewpoint's rotation is 0 degrees.
-    public init(viewpoint: Binding<Viewpoint?>, autoHide: Bool = true)
+    ///   - action: An action to perform when the compass is tapped.
+    public init(viewpoint: Binding<Viewpoint?>, action: (() async -> Void)? = nil)
```
Comment (dfeinzimer) [inline Documentation/Compass/README.md:47]:
```suggestion
    public init(viewpoint: Binding<Viewpoint?>, action: (() -> Void)? = nil)
```

**Review by dfeinzimer — COMMENTED** [review]
Looking good, I think it should be good to go after these last few changes.

**Review by dfeinzimer — APPROVED** [review]
Looks good!

File: Documentation/Compass/README.md [hunk 13]
```diff
@@ -35,27 +35,26 @@ Compass:
...
-`Compass` has the following modifier:
+`Compass` has the following modifiers:
 
 - `func compassSize(size: CGFloat)` - The size of the `Compass`, specifying both the width and height of the compass.
+- `func automaticallyHides(_:)` - Specifies whether the ``Compass`` should automatically hide when the heading is 0.
 
 ## Behavior:
 
-Whenever the map is not orientated North (non-zero bearing) the compass appears. When reset to north, it disappears. An initializer argument allows you to disable the auto-hide feature so that it always appears.
+Whenever the map is not orientated North (non-zero bearing) the compass appears. When reset to north, it disappears. The `automaticallyHides` view modifier allows you to disable the auto-hide feature so that it is always displayed.
```
Comment (rolson) [inline Documentation/Compass/README.md:57]: Should this then be named more like how SwiftUI names modifiers like this? 

maybe `disableAutoHide(:)`?

image

File: Sources/ArcGISToolkit/Components/Compass/Compass.swift [hunk 14]
```diff
@@ -41,14 +44,13 @@ public struct Compass: View {
     /// direction toward true East, etc.).
     /// - Parameters:
     ///   - heading: The heading of the compass.
-    ///   - autoHide: A Boolean value that determines whether the compass
-    ///   automatically hides itself when the heading is `0`.
+    ///   - action: An action to perform when the compass is tapped.
```
Comment (rolson) [inline Sources/ArcGISToolkit/Components/Compass/Compass.swift:47]: some doc about when you need to specify this action might be helpful to the user.  For example, does the compass set the heading to zero and call the action if I have an action set?  What is the behavior when no action is set?

File: Examples/Examples/CompassExampleView.swift [hunk 15]
```diff
@@ -15,27 +15,121 @@ import ArcGIS
...
+            MapView(map: map, viewpoint: viewpoint)
+                .onViewpointChanged(kind: .centerAndScale) { viewpoint = $0 }
+                .overlay(alignment: .topTrailing) {
+                    Compass(viewpoint: $viewpoint) {
+                        guard let viewpoint else { return }
+                        Task {
+                            // Animate the map view to zero when the compass is tapped.
+                            await proxy.setViewpoint(
+                                viewpoint.withRotation(0),
+                                duration: 0.25
```
[the remark this answers is not in the document]
Comment (mhdostal) [reply Examples/Examples/CompassExampleView.swift:81]: It still draws the compass, which is non-trivial, automatically allows it to track map view rotation, and will allow tapping to perform an action. The default is to rotate the map view to zero, but without animation, as that requires a `MapViewProxy`. So yea, there is value in it, but we can re-evaluate the default behavior to see if we can include animation in a seamless way.

File: Sources/ArcGISToolkit/Components/Compass/Compass.swift (hunk 14 above)
Comment (mhdostal) [reply Sources/ArcGISToolkit/Components/Compass/Compass.swift:47]: The default is to rotate the map view to zero, but without animation.

We'll update the doc.

File: Examples/Examples/CompassExampleView.swift (hunk 15 above)
[the remark this answers is not in the document]
Comment (rolson) [reply Examples/Examples/CompassExampleView.swift:81]: Good points. Thank you.

File: Documentation/Compass/README.md (hunk 13 above)
Comment (mhdostal) [reply Documentation/Compass/README.md:57]: That would be OK.  @dfeinzimer?

Comment (dfeinzimer) [reply Documentation/Compass/README.md:57]: https://github.com/Esri/arcgis-maps-sdk-swift-toolkit/pull/291

`disableAutocorrection` was deprecated in favor of `autocorrectionDisabled` so I'll follow the new scheme. e.g. `autoHideDisabled`.

**Pull request merged**